### PR TITLE
finish initial support for probabilistic forecast reports

### LINKED ISSUE #266: refactoring for probabilistic forecasts in reports
We need to extend/refactor a handful of things for probabilistic forecasts. Let's keep things simple and restrict reports to only probabilistic or only deterministic.

A to do list that is almost certainly incomplete:

- [x] Update report json metadata schema and `api.process_report_dict` parsing to allow for probabilistic forecasts. See #425 
- ~`ForecastObservation` and `ForecastAggregate` need to be modified or copied for probabilistic forecasts. They currently have an attribute `forecast: Forecast`. Could be `forecast: Union[Forecast, ProbabilisticForecast, ProbabilisticForecastConstantValue]`. Trivial.~ `ProbabilisticForecast, ProbabilisticForecastConstantValue` are `Forecast`, so I don't think the attribute spec needs to change.
- [x] `reports.main.get_data_for_report` assumes deterministic forecast. Trivial since `api.get_values` now understands probabilistic forecast types.
- [x] `api.get_values` allows for `ProbabilisticForecastConstantValues` but not `ProbabilisticForecast`. Consider modifying. In fact, there is no convenience method to get all forecast values for all `ProbabilisticForecastConstantValues` within a `ProbabilisticForecast`. Probably will want that in `api`.
- [x] `metrics.preprocessing` needs to support probabilistic forecasts. (it might already but it's not tested)
- [x] `solarforecastarbiter/metrics/calculator.py` (almost complete in #224) needs to support probabilistic forecasts
- [x] modify or create new template for probabilistic forecast reports. Modification probably gets us there for initial report.
- [x] report forecast links need to go to `/cdf/` or `/cdf/single/`
- [x] simple plot with lines for each `ProbabilisticForecastConstantValues` within a report. Might require labels like in https://github.com/SolarArbiter/solarforecastarbiter-dashboard/pull/238/files#diff-ec34245434455b3669a7ee1e5f3187feR894-R898 
- [x] same as above, but for each `ProbabilisticForecastConstantValues` within a `ProbabilisticForecast`.
- [x] simple plot with lines for `ProbabilisticForecastConstantValues` with `axis='x'`. Requires modification to at least plot y label.

related but address in separate issues:

- plot with progressively shaded areas for matched probability bands (e.g. 0/100, 10/90, 20/80, etc. percentiles)
- reliability diagram
- rank histogram

let me know what I'm missing.

@alorenzo175 any opinions before I get started?

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #266  .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Reports work with `ProbabilisticForecast` and `ProbabilisticForecastConstantValue` with `axis='y'`, though they are not pretty. Smaller legend text would help.

I think, but haven't tested, that reports will work with `axis='x'` but the plots will be mislabeled.

I've not fixed the links in the pdf report.

### CODE REVIEW

**Comment by wholmgren** [comment]
This [gist](https://gist.github.com/wholmgren/dee5ba8bee170a2eee1177d3f3ade0c8) shows a time series plot with a mix of `ProbabilisticForecast` and `ProbabilisticForecastConstantValue` on both the x and y axis. 

The y axis lines are good for `ProbabilisticForecastConstantValue` but a `ProbabilisticForecast` should probably have filled areas. That would require propagating even more metadata in `timeseries_meta_df`. An alternative might be building different `timeseries_value_df, timeseries_meta_df` for each set of obs, `ProbabilisticForecastConstantValue`, and `ProbabilisticForecats` in a report, and then passing those to `plotly_figures.timeseries` with a few more kwargs. In any case, it also requires fighting plotly's fill feature so I'm -1 on it for initial implementation.

Maybe (likely) there is a better way plot the x forecasts on the time series plot, but I can't think of it. Maybe we should stop trying to cram it into a single plot and we need a separate plot for probability vs time. This would require modifications to `plotly_figures.timeseries_plots` and the templates to make the 3rd plot optional.

**Comment by wholmgren** [comment]
@alorenzo175 feedback on the possible directions of this PR would be useful at this point.

**Comment by alorenzo175** [comment]
That plot is very hard for me to understand. Definitely would be helpful to separate the 'x' axis ones into a probability vs time

**Comment by wholmgren** [comment]
PR now separates plots into obs/deterministic/y-axis-prob

newplot(21)

and x-axis-prob

newplot(20)

**Comment by wholmgren** [comment]
@alorenzo175 this is ready for a review. I'm avoiding modifications to the pdf reports until the html is figured out.

**Comment by wholmgren** [comment]
The [gist](https://gist.github.com/wholmgren/dee5ba8bee170a2eee1177d3f3ade0c8#file-test_xy_prob-html) now has an html file with the rendered report for a set of forecasts that includes axis = x and y.

Screen Shot 2020-06-24 at 2 14 09 PM

File: solarforecastarbiter/reports/template.py [hunk 1]
```diff
@@ -89,6 +89,8 @@ def _get_render_kwargs(report, dash_url, with_timeseries):
             timeseries_specs = ('{}', '{}')
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/template.py:89]: need to add a 3rd element for [2] below

**Comment by wholmgren** [comment]
PDF links now work (tested locally).

I added a couple of tests to `test_cli.py` for probabilistic forecasts. The first has only forecasts on y axis - this makes a report with just the variable vs time plot and the scatter plot. The second has forecasts on both x and y - this adds the probability vs time plot. I confirmed that they render correctly by inspecting the output of the test functions. I could add a test to `test_plotly_figures` for the json/None behavior of `timeseries` but that function is not currently explicitly tested, probably because it relies on a report object that is a pain to construct.

The untested line in the diff was not tested before this PR.

**Comment by wholmgren** [comment]
Added a new test for `timeseries_plots` to check that `str, str, None` is returned for the `conftest.raw_report`. Do we need a test for `str, str, str` based on a new `conftest.raw_report_xy` or some other method? Also added a new test for the naming function.

File: solarforecastarbiter/reports/templates/pdf/macros.j2 [hunk 2]
```diff
@@ -115,7 +115,7 @@
     \bottomrule
     \endlastfoot
     %- for pfxobs in report.raw_report.processed_forecasts_observations
-    \VAR{pfxobs.name} &
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")} &
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/templates/pdf/macros.j2:118]: why not use the html_to_tex filter?

File: solarforecastarbiter/reports/templates/pdf/macros.j2 [hunk 3]
```diff
@@ -125,11 +125,23 @@
     %- endif
     \VAR{pfxobs.original.data_object.interval_label} &
     \VAR{pfxobs.original.data_object.interval_length.total_seconds()//60 | int} min &
+    %- if pfxobs.original.forecast.constant_values is defined
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/templates/pdf/macros.j2:128]: macro?

Comment (alorenzo175) [reply solarforecastarbiter/reports/templates/pdf/macros.j2:128]: Oh I guess i don't know about macro within macro

File: solarforecastarbiter/reports/templates/pdf/macros.j2 [hunk 4]
```diff
@@ -187,7 +199,7 @@
     \bottomrule
     \endlastfoot
     %- for pfxobs in proc_fxobs_list
-    \VAR{pfxobs.name} & \VAR{pfxobs.original.data_object.name}
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")} & \VAR{pfxobs.original.data_object.name}
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/templates/pdf/macros.j2:202]: filter?

File: solarforecastarbiter/reports/templates/pdf/macros.j2 [hunk 5]
```diff
@@ -245,7 +257,7 @@
     \bottomrule
     \endlastfoot
     %- for pfxobs in proc_fxobs_list
-    \VAR{pfxobs.name}
+    \VAR{pfxobs.name | replace("<=", "$\leq$") | replace("%", "\%")}
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/templates/pdf/macros.j2:260]: filter?

File: solarforecastarbiter/tests/test_cli.py [hunk 6]
```diff
@@ -328,6 +328,45 @@ def test_report_pdf(cli_token, mocker, report_objects, remove_orca):
...
+    mocker.patch('solarforecastarbiter.cli.APISession.process_report_dict',
+                 return_value=report.replace(status=''))
+    mocker.patch('solarforecastarbiter.cli.reports.get_data_for_report',
+                 return_value=cdf_and_cv_report_data)
+    runner = CliRunner()
+    with tempfile.TemporaryDirectory() as tmpdir:
+        infile = tmpdir + '/report.json'
+        with open(infile, 'w') as f:
+            f.write('{}')
+        outfile = tmpdir + '/test_out.html'
```
Comment (alorenzo175) [inline solarforecastarbiter/tests/test_cli.py:344]: do both pdf and html?

**Comment by alorenzo175** [comment]
> Added a new test for `timeseries_plots` to check that `str, str, None` is returned for the `conftest.raw_report`. Do we need a test for `str, str, str` based on a new `conftest.raw_report_xy` or some other method? 

I guess it wouldn't hurt?

File: solarforecastarbiter/reports/templates/pdf/macros.j2 (hunk 2 above)
Comment (wholmgren) [reply solarforecastarbiter/reports/templates/pdf/macros.j2:118]: ah, I did not understand until now that it's made accessible for calling within a jinja variable expression.

File: solarforecastarbiter/reports/templates/pdf/macros.j2 (hunk 3 above)
Comment (wholmgren) [reply solarforecastarbiter/reports/templates/pdf/macros.j2:128]: I looked at it for a bit but had a hard time with it

**Comment by wholmgren** [comment]
@alorenzo175 here's the last test. this copy pasta pairs nicely with an IPA.

**Comment by alorenzo175** [comment]
good with me
